zerver/tests/test_auth_backends.py

```diff
@@ -7212,5 +7216,156 @@ def test_invalid_realm_for_user_failure(self) -> None:
             self.assert_json_error_contains(result, "Invalid subdomain", 404)
 
 
+class LDAPGroupSyncTest(ZulipTestCase):
+    @override_settings(AUTHENTICATION_BACKENDS=("zproject.backends.ZulipLDAPAuthBackend",))
+    def test_ldap_group_sync(self) -> None:
+        self.init_default_ldap_database()
+
+        with self.settings(
+            LDAP_GROUP_SYNC_LIST={
+                "zulip": [
+                    "ldap_group_not_existing_in_zulip",
+                ]
+            }
+        ), self.assertLogs("zulip.ldap", "DEBUG") as ldap_log:
+            ensure_user_groups_for_ldap_sync_exist()
+            self.assertTrue(
+                UserGroup.objects.filter(name="ldap_group_not_existing_in_zulip").exists()
+            )
+
+        self.assertEqual(
+            ldap_log.output,
+            [
+                "DEBUG:zulip.ldap:creating zulip group ldap_group_not_existing_in_zulip from ldap for realm zulip"
+            ],
+        )
+
+        user_profile = self.example_user("hamlet")
+        with self.settings(LDAP_APPEND_DOMAIN="zulip.com"):
+            result = sync_user_from_ldap(user_profile, mock.Mock())
+            self.assertTrue(result)
+        self.assertTrue(user_profile.is_active)
+
+        with self.settings(
+            AUTH_LDAP_GROUP_SEARCH=LDAPSearch(
+                "ou=groups,dc=zulip,dc=com",
+                ldap.SCOPE_ONELEVEL,
+                "(objectClass=groupOfUniqueNames)",
+            ),
+            LDAP_GROUP_SYNC_LIST={
+                "zulip": [
+                    "cool_test_group",
+                ]
+            },
+            LDAP_APPEND_DOMAIN="zulip.com",
+        ), self.assertLogs("zulip.ldap", "DEBUG") as zulip_ldap_log:
+            self.assertFalse(
+                UserGroup.objects.filter(realm=get_realm("zulip"), name="cool_test_group").exists()
+            )
+
+            # Create group cool_test_group since it doesn't yet exist in Zulip
+            ensure_user_groups_for_ldap_sync_exist()
+            # Run the function a second time to test for bugs, this should noop at this point.
+            ensure_user_groups_for_ldap_sync_exist()
+
+            self.assertTrue(
+                UserGroup.objects.filter(realm=get_realm("zulip"), name="cool_test_group").exists()
+            )
+
+            user_group = UserGroup.objects.get(realm=get_realm("zulip"), name="cool_test_group")
```

Comment:
As a test style thing, we should do `realm=get_realm("zulip")` once at the top of the test function, rather than repeating it many times in the test.

Reply:
see eb37e19e9a05c5dc656ef419b53e5a7c8b8b6355